Added buttons on sounds explorer window to mute all attached / rezzed sounds by avatar

Gaalrauch opened the pull request:
Added buttons on sounds explorer window to mute all attached / rezzed sounds by avatar

![Screen1](https://github.com/user-attachments/assets/addf53e5-459d-4041-b71b-59de514f202b)
![Screen2](https://github.com/user-attachments/assets/9ccba154-faf8-40b8-b2f9-b68f101bf879)


Just in case: I did not update XML files for all langages, only EN

(sorry for multiple posts, I'm not a git master)

Gaalrauch commented:
Hi, thanks. I'm sorry about all those "random" lines and in fact I forget to remove a temp comment I made in french xD. I made a version with 3 types, attached, rezzed, and gesture sounds. You may be interested by this one ?

Ansariel commented:
I am trying to find the actual usecase for this, because:

- If you find a sound annoying in general, you probably would block it in general - or are they situations in which you like to hear that sound?
- On the other hand, if it is a normal sound but somebody is walking around spamming it, it's most likely a troll/griefer and you would block that person anyway.

This seems to be a very niche usecase (if it exists) and introducing 3 new asset types just for that seems overkill - not to mention the problem that arises if LL adds a new asset type as well, which then results in having to change the asset type for these blacklist asset types, which also means there would be a conversion of the saved blacklisted assets.

Gaalrauch commented:
For now, I’ve implemented eBlacklistFlag as: 

    enum class eBlacklistFlag
    {
        NONE = 0,
        WORN,
        REZZED,
        GESTURE
    };
 
This allows the same avatar to appear up to three times in the blacklist window, with three different types, like this: 
![image](https://github.com/user-attachments/assets/06b44e8f-6804-414a-ac45-0e7157f0236a)

Currently, removing one of these three entries also removes the others.

I haven’t yet implemented combined values (bitflags) as suggested by Ansariel’s original design of eBlacklistFlag, but I might revisit that depending on how useful or necessary it becomes.

Ansariel commented on indra/newview/NACLfloaterexploresounds.cpp:
```diff
@@ -513,3 +519,159 @@ void NACLFloaterExploreSounds::onBlacklistAvatarNameCacheCallback(const LLUUID&
 }
+
+// add avatar attachments sounds to blacklist
+void NACLFloaterExploreSounds::blacklistAvatarSoundsByFlagWorn()
```
You only need 2 methods here in total:

- blacklistAvatarSoundsByFlag
- onBlacklistAvatarSoundsByFlagNameCacheCallback

The flag enum value can be passed in/through in the boost::bind calls.

Ansariel commented on indra/newview/fsassetblacklist.h:
```diff
@@ -46,11 +46,20 @@ class FSAssetBlacklist : public LLSingleton<FSAssetBlacklist>
     void removeItemFromBlacklist(const LLUUID& id);
     void removeItemsFromBlacklist(const uuid_vec_t& ids);
     void saveBlacklist();
+    enum class eBlacklistFlag
```
As mentioned in my previous comment, I would prefer having the enum setup as bitfield so you can have multiple combinations of flags set/checked at the same time.

Gaalrauch commented:
Merged the previously separate blacklist handling for WORN, REZZED, and GESTURE into a unified system using flags.
Refactored eBlacklistFlag to behave as a bitfield enum, enabling combinations like WORN | GESTURE.
Updated existing blacklist-related functions (such as isBlacklisted(), addNewItemToBlacklist(), etc.) to accept an eBlacklistFlag parameter.
Implemented flag-aware logic in isBlacklisted() so that checks only return true if the specified flag(s) are present.
Modified addNewItemToBlacklist() to:
    Add a new flag to an existing entry if the UUID is already blacklisted.
    Create a new entry if none exists.
Improved removeFlagFromItem() to remove individual flags while keeping the remaining ones intact, and delete the entry only when no flags remain.
Adjusted the UI to display one list entry per active flag.
Ensured list updates are clean and avoid duplication by:
    Removing existing entries before re-adding updated flag sets.
    Deferring reconstruction when multiple flags are being removed in batch.

Ansariel commented:
This looks already better. It probably would be better to just add another column which shows the flag value instead of reusing the "type" column (the long texts don't fit in there by default anyway).

My original idea was to have one entry for the asset and then have a column for the flags, showing them all in just a single string. That however would have required a bunch of extra logic to just remove one of the flags, so... Maybe we leave that for now. However, adding a new column which shows the label for each flag would be nice instead of re-using the "type" column. Then you don't need that huge chunk of duplicated code in `FSFloaterAssetBlacklist::addElementToList`

Other than that it looks fine to me.

Ansariel commented on indra/newview/fsassetblacklist.cpp:
```diff
@@ -87,29 +87,72 @@ bool FSAssetBlacklist::isBlacklisted(const LLUUID& id, LLAssetType::EType type)
+    }
+    else
+    {
+        // Nouvelle entr�e
```
English comment please

Gaalrauch replied:
Oops

Gaalrauch commented:
I'm wondering whether it might be better to replace the use of AT_SOUND with AT_PERSON for these features. We're not really muting a sound itself, but rather something that belongs to an avatar. Also, should we consider renaming the flags if we’re planning for additional use cases? For example:

enum class eBlacklistFlag
{
    NONE    = 0,
    MUTE_AVATAR_WORN_OBJECTS_SOUNDS    = 1 << 0,
    MUTE_AVATAR_REZZED_OBJECTS_SOUNDS  = 1 << 1,
    MUTE_AVATAR_GESTURES_SOUNDS        = 1 << 2
};

Ansariel commented:
You are blacklisting a sound - blacklisting a person means not rendering it at all.

Could you please address this:

> It probably would be better to just add another column which shows the flag value instead of reusing the "type" column (the long texts don't fit in there by default anyway). That makes it more general for possible future extensions.

Gaalrauch commented:
Done, I removed the "huge chunk of duplicated code" in FSFloaterAssetBlacklist::addElementToList too

Ansariel commented:
I'll merge this - then we'll see how it goes.

chanayane commented:
Nevermind, a full rebuild fixed it. Sorry.
